### PR TITLE
contact: replace ps system call with psutil

### LINKED ISSUE #3965: evaluate the detect_old_contact_file logic
This logic could do with a quick check before 8.0.0 to make sure it is properly Cylc8'ed.

See https://github.com/cylc/cylc-flow/pull/3956#issuecomment-734224421

It's purpose is to tell whether the scheduler process still running or not irrespective of whether we are able to contact it via TCP or not. If we are 100% certain that it's not running then we can remove the contact file.

This logic was implemented in older Cylc, we should check it over in Cylc8 to make sure it is still watertight, especially with command reinvocation (i.e. `cylc run --host`).

### LINKED ISSUE #4416: Cylc Flow ps command syntax not compatible with Alpine Linux busybox
**Describe the bug**

Running a workflow with Cylc Flow 8.0b2 will result in a `ps` command being executed that is not compatible with the [Busybox](https://www.busybox.net/) syntax. Something changed in the syntax from 8.0a1 as that version still worked with Alpine linux.

```bash
2021-09-17T00:20:18Z INFO - Final point: None
2021-09-17T00:20:18Z ERROR - cannot get process "args" from "ps": ps: unrecognized option: w
	BusyBox v1.33.1 () multi-call binary.
	Usage: ps [-o COL1,COL2=HEADER] [-T]
	Show list of processes
	        -o COL1,COL2=HEADER     Select columns for display
	        -T                      Show threads
	Traceback (most recent call last):
	  File "/usr/local/lib/python3.7/site-packages/cylc/flow/scheduler.py", line 562, in start_scheduler
	    self._configure_contact()
	  File "/usr/local/lib/python3.7/site-packages/cylc/flow/scheduler.py", line 1009, in _configure_contact
	    'cannot get process "args" from "ps": %s' % err)
	RuntimeError: cannot get process "args" from "ps": ps: unrecognized option: w
	BusyBox v1.33.1 () multi-call binary.
	Usage: ps [-o COL1,COL2=HEADER] [-T]
	Show list of processes
	        -o COL1,COL2=HEADER     Select columns for display
	        -T                      Show threads
	
2021-09-17T00:20:18Z CRITICAL - Workflow shutting down - cannot get process "args" from "ps": ps: unrecognized option: w
	BusyBox v1.33.1 () multi-call binary.
	Usage: ps [-o COL1,COL2=HEADER] [-T]
	Show list of processes
	        -o COL1,COL2=HEADER     Select columns for display
	        -T                      Show threads
	
2021-09-17T00:20:19Z INFO - DONE

```

**Release version(s) and/or repository branch(es) affected?**

8.0b2.

**Steps to reproduce the bug**

- clone <https://github.com/kinow/cylc-docker/tree/>
- `cd cylc-flow/8.0b2`
- `docker build --tag kinow/cylc-flow:8.0b2-alpine .`
- `docker run --rm --name cylc-flow -t -i kinow/cylc-flow:8.0b2-alpine`

Follow instructions from the image `README.md` file, or just install and play any workflow. You should see the error message in the workflow logs.

**Expected behavior**

No errors.

**Screenshots**

**Additional context**

Not a blocker for releases, but in case there are no plans to support busybox syntax, it might be good to document that cylc is not compatible with alpine distro (and any others that use it). Debian has a slim image that's small enough too.

**Pull requests welcome!**
This is an Open Source project - please consider contributing a bug fix
yourself (please read `CONTRIBUTING.md` before starting any work though).

### PR DESCRIPTION
**Pull request by oliver-sanders**
Closes #4416
Closes #3965

* Replaces the compound pid:command contact file fields with separate
  pid & contact fields.
* Replaces system `ps` call with Python `psutil`.
* Fixes a bug reported on Alpine Linux & improves portability.
* Reviewed and updated the detection of stale contact files.
* Added tests.

**Requirements check-list**
- [x] I have read `CONTRIBUTING.md` and added my name as a Code Contributor.
- [x] Contains logically grouped changes (else tidy your branch by rebase).
- [x] Does not contain off-topic changes (use other PRs for other changes).
- [x] Applied any dependency changes to both `setup.py` and
  `conda-environment.yml`.
- [x] Appropriate tests are included (unit and/or functional).
- [x] Appropriate change log entry included.
- [x] No documentation update required.

### CODE REVIEW

File: cylc/flow/workflow_files.py [hunk 1]
```diff
@@ -336,85 +349,148 @@ class RemoteCleanQueueTuple(NamedTuple):
...
+        # doesn't match
+        >>> _is_process_running('localhost', proc.pid, ['something-else'])
+        False
+
+    """
+    # See if the process is still running or not.
+    cmd = ['cylc', 'psutil']
+    metric = f'[["Process", {pid}]]'
+    if is_remote_host(host):
+        cmd = _construct_ssh_cmd(cmd, host)
```
Comment (oliver-sanders) [inline cylc/flow/workflow_files.py:393]: > Note: `construct_ssh_cmd` takes `platform` as an argument and respects the platform config (incl SSH settings) whereas `_construct_ssh_cmd` takes `host` as an argument and uses the config for `localhost`.
>
> We were doing `construct_ssh_cmd(get_platform())` before which is effectively the same.

File: cylc/flow/daemonize.py [hunk 2]
```diff
@@ -98,7 +97,6 @@ def daemonize(schd):
                 "host": schd.host,
                 "url": workflow_url,
                 "pub_url": pub_url,
-                "ps_opts": PS_OPTS,
```
Comment (oliver-sanders) [inline cylc/flow/daemonize.py:101]: Removed one field from the daemonization output.

File: cylc/flow/scheduler.py [hunk 3]
```diff
@@ -1017,12 +1003,14 @@ def _configure_contact(self):
...
                 self.owner,
             fields.PORT:
-                str(self.server.port),
-            fields.PROCESS:
-                process_str,
+                str(self.server.port),  # type: ignore
+            fields.PID:
+                str(proc.pid),
+            fields.COMMAND:
+                ' '.join(proc.cmdline()),
```
Comment (oliver-sanders) [inline cylc/flow/scheduler.py:1010]: Split the `PROCESS` contact file field into `PID` and `COMMAND`.

File: cylc/flow/scripts/psutil.py [hunk 4]
```diff
@@ -19,6 +19,11 @@
 (This command is for internal use.)
 
 For internal use with the `cylc.flow.host_select` module.
```
Comment (hjoliver) [inline cylc/flow/scripts/psutil.py:21]: and ...

File: cylc/flow/workflow_files.py [hunk 5]
```diff
@@ -262,9 +267,19 @@ class ContactFileFields:
...
     """The user account under which the scheduler process is running."""
 
-    PROCESS = 'CYLC_WORKFLOW_PROCESS'
+    PID = 'CYLC_WORKFLOW_PID'
     """The process ID of the running workflow on ``CYLC_WORKFLOW_HOST``."""
 
+    COMMAND = 'CYLC_WORKFLOW_COMMAND'
+    """The command that was used to run the workflow on ``CYLC_WORKFLOW_HOST```.
+
+    Note that this command may be effected by:
```
Comment (hjoliver) [inline cylc/flow/workflow_files.py:276]:
```suggestion
    Note that this command may be affected by:
```

File: cylc/flow/workflow_files.py [hunk 6]
```diff
@@ -336,85 +349,148 @@ class RemoteCleanQueueTuple(NamedTuple):
...
+    process = json.loads(out)[0]
+    return ' '.join(process['cmdline']) == command
+
+
+def detect_old_contact_file(reg: str, contact_data=None) -> None:
+    """Check if the workflow process is still running.
+
+    As a side-effect this should detect and rectify the situation
+    where an old contact file is still present from a previous run. This can be
+    caused by the uncontrolled teardown or a running Scheduler (e.g. a power
```
Comment (hjoliver) [inline cylc/flow/workflow_files.py:427]: or?
```suggestion
    caused by the uncontrolled teardown of a running Scheduler (e.g. a power
```

**Review by hjoliver — COMMENTED** [review]
From a quick look through, looks good, and much clearer than the old code :+1:

File: cylc/flow/workflow_files.py [hunk 7]
```diff
@@ -262,9 +268,19 @@ class ContactFileFields:
     OWNER = 'CYLC_WORKFLOW_OWNER'
     """The user account under which the scheduler process is running."""
 
-    PROCESS = 'CYLC_WORKFLOW_PROCESS'
+    PID = 'CYLC_WORKFLOW_PID'
     """The process ID of the running workflow on ``CYLC_WORKFLOW_HOST``."""
 
+    COMMAND = 'CYLC_WORKFLOW_COMMAND'
+    """The command that was used to run the workflow on ``CYLC_WORKFLOW_HOST```.
```
Comment (kinow) [inline cylc/flow/workflow_files.py:275]: `CYLC_WORKFLOW_HOST` has one more quote on its right hand side?

**Review by kinow — COMMENTED** [review]
Code looks good. I'm on Windows at the moment, so will test tomorrow with that alpine container from my Linux box 👍 thanks!

**Comment by oliver-sanders** [comment]
Integration test is failing on Mac OS (GitHub), putting this on hold until resolved.

**Comment by oliver-sanders** [comment]
The MacOS failure was related to the parallelism of tests on the GH runner, the PID test I had written was flagging false positives which couldn't be safely fixed so I have removed it.

**Comment by kinow** [comment]
> The MacOS failure was related to the parallelism of tests on the GH runner, the PID test I had written was flagging false positives which couldn't be safely fixed so I have removed it.

A GH macos job is still failing :cold_sweat:

**Comment by oliver-sanders** [comment]
Different test thankfully, should pass on rerun.

**Review by kinow — APPROVED** [review]
Tested on [Alpine](https://github.com/kinow/cylc-docker/blob/c6324337a57e6f02117572732a2cfd76f1e61b62/cylc-flow/8.0b2/Dockerfile) container with this diff:

```diff
diff --git a/cylc-flow/8.0b2/Dockerfile b/cylc-flow/8.0b2/Dockerfile
index 34ce558..804c351 100644
--- a/cylc-flow/8.0b2/Dockerfile
+++ b/cylc-flow/8.0b2/Dockerfile
@@ -43,7 +43,7 @@ RUN apk update && \
     g++ \
     linux-headers \
     python3-dev && \
-    pip install --no-cache-dir "cylc-flow==${cylc_docker_version}" && \
+    pip install --no-cache-dir git+https://github.com/oliver-sanders/cylc-flow.git@7377b3cffe57ac01102cbc77e0ae669266bd95cf#egg=cylc-flow && \
     apk del build-dependencies
 
 # Add a non-root user
@@ -69,7 +69,7 @@ RUN chown -R ${USER_NAME}:${USER_GROUP} ${USER_ROOT} && \
 
 # FIXME: https://github.com/cylc/cylc-flow/issues/4416
 #        hacking due to incompatible ps args
-RUN sed -i 's/wopid/opid/' /usr/local/lib/python3.7/site-packages/cylc/flow/workflow_files.py
+# RUN sed -i 's/wopid/opid/' /usr/local/lib/python3.7/site-packages/cylc/flow/workflow_files.py

```

Workflow ran with no issues this time :tada: Thanks!

**Comment by hjoliver** [comment]
`tests/f/restart/00-pre-initial.t` failing here

File: cylc/flow/network/client.py [hunk 8]
```diff
@@ -272,13 +277,30 @@ def _timeout_handler(workflow: str, host: str, port: Union[int, str]):
...
+        contact_host: str = contact_data.get(ContactFileFields.HOST, '?')
+        contact_port: str = contact_data.get(ContactFileFields.PORT, '?')
+        if (
+            contact_host != host
+            or contact_port != str(port)
+        ):
+            raise CylcError(
+                f'The workflow is no longer running at {host}:{port}\n'
+                f'It has moved to {contact_host}:{contact_port}'
+            )
```
Comment (hjoliver) [inline cylc/flow/network/client.py:296]: This causes `cylc stop --poll` to fail if the target workflow got restarted quickly, hence failure of `tests/f/restart/00-pre-initial.t`.  I'll push a quick fix and merge the PR if it works.

Comment (hjoliver) [reply cylc/flow/network/client.py:296]: (The fix is to intercept CylcError in the poller, not to avoid raising it here).

**Review by hjoliver — APPROVED** [review]

**Pull request merged**